src/notifications/notificationmanager.cpp

```diff
@@ -121,14 +192,76 @@ QPair<QString, QString> processProperties(uint pid)
     return rv;
 }
 
-bool notificationReverseOrder(const LipstickNotification *lhs, const LipstickNotification *rhs)
+static bool notificationReverseOrder(const LipstickNotification *lhs, const LipstickNotification *rhs)
 {
     // Sort least significant notifications first
     return *rhs < *lhs;
 }
 
 }
 
+ClientIdentifier::ClientIdentifier(QObject *parent, const QDBusConnection &connection, const QDBusMessage &message)
+    : QObject(parent)
+    , m_connection(connection)
+    , m_message(message)
+    , m_clientPid(-1)
+{
+    QDBusMessage request = QDBusMessage::createMethodCall("org.freedesktop.DBus", "/org/freedesktop/DBus", "org.freedesktop.DBus", "GetConnectionUnixProcessID");
+    request << clientName();
+    NOTIFICATIONS_DEBUG("identify" << member() << "from" << clientName() << "...");
+    QDBusPendingReply<quint32> call = this->connection().asyncCall(request);
+    QDBusPendingCallWatcher *getPidWatcher = new QDBusPendingCallWatcher(call, this);
+    connect(getPidWatcher, &QDBusPendingCallWatcher::finished, this, &ClientIdentifier::getPidReply);
+}
+
+void ClientIdentifier::getPidReply(QDBusPendingCallWatcher *getPidWatcher)
+{
+    bool waiting = false;
+    QDBusPendingReply<quint32> reply = *getPidWatcher;
+    if (reply.isError()) {
+        qWarning() << "identify" << member() << "from" << clientName() << "- GetConnectionUnixProcessID:" << reply.error().name() << reply.error().message();
+    } else {
+        quint32 pid = reply.value();
+        if (pid > 0) {
+            m_clientPid = pid;
+            if (processIsDBusProxy(pid)) {
+                QDBusMessage request = QDBusMessage::createMethodCall(clientName(), "/", "org.sailfishos.sailjailed", "Identify");
+                QDBusPendingReply<QVariantMap> call = connection().asyncCall(request);
+                QDBusPendingCallWatcher *identifyWatcher = new QDBusPendingCallWatcher(call, this);
+                connect(identifyWatcher, &QDBusPendingCallWatcher::finished, this, &ClientIdentifier::identifyReply);
+                waiting = true;
+            }
+        }
+    }
+    if (!waiting)
+        finish();
+    getPidWatcher->deleteLater();
+}
+
+void ClientIdentifier::identifyReply(QDBusPendingCallWatcher *identifyWatcher) {
+    QDBusPendingReply<QVariantMap> reply = *identifyWatcher;
+    if (reply.isError()) {
+        qWarning() << "identify" << member() << "from" << clientName() << " - Identify:" << reply.error().name() << reply.error().message();
+    } else {
+        QVariantMap map(reply.value());
+        if (map.contains("pid")) {
+            bool ack = false;
+            int pid = map["pid"].toInt(&ack);
+            if (ack && pid > 0) {
+                m_clientPid = pid;
+            }
+        }
+    }
+    finish();
+    identifyWatcher->deleteLater();
+}
+
+void ClientIdentifier::finish()
+{
+    NOTIFICATIONS_DEBUG("identify" << member() << "from" << clientName() << "-> using pid" << clientPid());
+    Q_EMIT finished(this);
```

Comment:
Can consider also just emitting finished() directly where it's needed and using sender() for getting the pointer.

Reply:
hmm. I did not know about sender(). Using it would produce shorter / cleaner code. But also sort of hide the bit that this signal is also object ownership transfer. But unnecessary args are unnecessary. I guess I'll change it to use sender().